docs: containize instructions with `docker build` (#3113)
bentoml/BentoML: docs/source/guides/migration.rst

1.0 Migration Guide
===================

> BentoML version 1.0.0 APIs are backward incompatible with version 0.13.1. However, most of the common
> functionality can be achieved with the new version. We will guide and demonstrate the migration by
> transforming the `quickstart <https://github.com/bentoml/BentoML/tree/main/examples/quickstart>`_ gallery project

Reviewer: Will probably need a reflow. (I know this is a just a whitespace change, but we may as well make these changes)

```suggestion
BentoML version 1.0.0 APIs are backward incompatible with version 0.13.1. However, most of the
functionality is maintained in the new version. We will demonstrate this by
migrating the `quickstart <https://github.com/bentoml/BentoML/tree/main/examples/quickstart>`_ gallery project.
```

.. admonition:: 💡 Migration Task

   Update the service definition module ``service.py`` from the BentoML 0.13.1 specification to 1.0.0 specification.

.. tab-set::

    .. tab-item:: 0.13.1

       .. code-block:: python
          :caption: service.py

          import pandas as pd

          from bentoml import env, artifacts, api, BentoService
          from bentoml.adapters import DataframeInput
          from bentoml.frameworks.sklearn import SklearnModelArtifact

          @env(infer_pip_packages=True)
          @artifacts([SklearnModelArtifact('model')])
          class IrisClassifier(BentoService):
              @api(input=DataframeInput(), batch=True)
              def predict(self, df: pd.DataFrame):
                  return self.artifacts.model.predict(df)

    .. tab-item:: 1.0.0

       .. code-block:: python
          :caption: service.py

          import numpy as np
          import pandas as pd

          import bentoml

          iris_clf_runner = bentoml.sklearn.get("iris_clf:latest").to_runner()

          svc = bentoml.Service("iris_classifier", runners=[iris_clf_runner])

>           @svc.api(input=bentoml.io.PandasDataFrame(), output=bentoml.io.NumpyNdarray())

Reviewer: Why this change? It's not in the actual example.

Author: let me change this back

    2022-07-01T21:57:48+0000 [INFO] [api_server:3] Service loaded from Bento directory: bentoml.Service(tag="iris_classifier:6otbsmxzq6lwbgxi", path="/home/bentoml/bento/")
    2022-07-01T21:57:48+0000 [INFO] [api_server:4] Service loaded from Bento directory: bentoml.Service(tag="iris_classifier:6otbsmxzq6lwbgxi", path="/home/bentoml/bento/")

.. note::

>    BentoML 1.x leverage BuildKit <https://github.com/moby/buildkit>`_, to

Reviewer: ```suggestion
   BentoML 1.x leverages BuildKit <https://github.com/moby/buildkit>`_, to
```

   containerize Bentos.

   BuildKit has been `integrated into Docker 18.09 and higher <https://docs.docker.com/develop/develop-images/build_enhancements/>`_.
   Providing the environment variable ``DOCKER_BUILDKIT=1`` will enable BuildKit:

   .. code-block:: bash

>       $ DOCKER_BUILDKIT=1 bentoml containerize iris_classifier:latest

Reviewer: Can we automatically set this in our CI and not require `buildx`?

Author: Our containerize already set this environment variable.

.. seealso::

>    :ref:`Containerize Bentos <concepts/bento:Docker Options>` to learn more about our
>    docker options and how to customize the docker image. :ref:`guides/containerization:Advanced Containerization` goes into more 
>    details about advanced containerization features that BentoML provides, for those who
>    are interested.

Reviewer: ```suggestion
   :ref:`concepts/bento:Docker Options` describes our
   docker options and how to customize the docker image.
   :ref:`guides/containerization:Advanced Containerization` goes into more 
   details about advanced containerization features that BentoML provides.
```

   By providing the environment variable ``DOCKER_BUILDKIT=1``, Docker will use BuildKit
   to containerize your Bento.

   If you are using :ref:`bentoml containerize <reference/cli:containerize>`, BentoML
   will automatically set the environment variable ``DOCKER_BUILDKIT=1`` for you:

   .. code-block:: bash

      $ bentoml containerize iris_classifier:latest

   If you are using ``docker build``, you will need to set the environment variable
   ``DOCKER_BUILDKIT=1`` yourself. Make sure to also pass in the build context path to
   be the Bento path. See `docker build documentation <https://docs.docker.com/engine/reference/commandline/build/#usage>`_ for more details.

   .. code-block:: bash

      $ BENTO_PATH=$(bentoml get iris_classifier:latest -o path)
      $ DOCKER_BUILDKIT=1 docker build -t iris_classifier:v1 \
                                       -f $BENTO_PATH/env/docker/Dockerfile \
                                       $BENTO_PATH

   :bdg-info:`Remarks:` We recommend users to **always** use :ref:`bentoml containerize <reference/cli:containerize>` to containerize Bentos.

.. seealso::

>    :ref:`Containerize Bentos <concepts/bento:Docker Options>` to learn more about our
>    docker options and how to customize the docker image.
> 
>    :ref:`guides/containerization:Advanced Containerization` goes into more 
>    details about advanced containerization features that BentoML provides, for those who
>    are interested.

Author: ```suggestion
   :ref:`Containerize Bentos <concepts/bento:Docker Options>` describes our
   docker options and how to customize the docker image.

   :ref:`guides/containerization:Advanced Containerization` goes into more 
   details about the advanced containerization features that BentoML provides.
```
